File: jvm/periph-java/src/main/java/it/uhde/periph/chips/led/WS2812BFull.java
package it.uhde.periph.chips.led;

import it.uhde.periph.connection.Connection;

/**
 * WS2812B full interface — extends {@link WS2812BMinimal} with per-pixel control.
 *
 * <p>Adds individual pixel addressing, explicit {@link #show()}, global brightness
 * scaling (0–255), pixel buffer rotation, and HSV fill. Call {@link #setPixel}
 * or {@link #setPixels} to update the buffer, then {@link #show()} to transmit.
 * The inherited {@link #fill} remains available as the fast path for
 * all-same-colour updates (fills and transmits immediately).
 *
 * <p>Brightness is stored separately and applied non-destructively at
 * {@link #show()} time: {@code sent = stored × brightness / 255}.
 */
public class WS2812BFull extends NeoPixelRGBFull {

    private static final int[] CHANNEL_ORDER = {1, 0, 2}; // GRB: wire[0]=G, wire[1]=R, wire[2]=B
    private static final int RESET_BYTES = 16;             // ~53us, WS2812B's default minimum

    /**
     * Construct the full driver.
     *
     * @param connection configured NeoPixel connection
     * @param n         number of pixels in the strip (≥1)
     */
    public WS2812BFull(Connection connection, int n) {
        super(connection, n, CHANNEL_ORDER, RESET_BYTES);
    }
}

File: jvm/periph-kotlin/src/main/kotlin/it/uhde/periph/chips/led/WS2812BFull.kt
package it.uhde.periph.chips.led

import it.uhde.periph.connection.Connection

/**
 * WS2812B full interface — extends the shared [NeoPixelRGBFull] Java base
 * with GRB wire order and WS2812B's default reset length.
 *
 * Adds individual pixel addressing, explicit `show`, global brightness
 * scaling (0–255), pixel buffer rotation, and HSV fill. Call `setPixel`
 * or `setPixels` to update the buffer, then `show` to transmit.
 * The inherited `fill` remains available as the fast path for
 * all-same-colour updates (fills and transmits immediately).
 *
 * @param connection configured NeoPixel connection
 * @param n number of pixels in the strip (≥1)
 */
class WS2812BFull(connection: Connection, n: Int) :
    NeoPixelRGBFull(connection, n, CHANNEL_ORDER, RESET_BYTES) {

    companion object {
        private val CHANNEL_ORDER = intArrayOf(1, 0, 2) // GRB: wire[0]=G, wire[1]=R, wire[2]=B
        private const val RESET_BYTES = 16               // ~53us, WS2812B's default minimum
    }
}
